File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_py/EightPuzzle.py
import collections
import copy

class EightPuzzle:

    def __init__(self, initialState):
        self.initial_state = [row[:] for row in initialState]
        self.goal_state = ((1, 2, 3), (4, 5, 6), (7, 8, 0))

    def find_blank(self, state):
        for i in range(3):
            for j in range(3):
                if state[i][j] == 0:
                    return (i, j)
        return None

    def move(self, state, direction):
        i, j = self.find_blank(state)
        new_state = [row[:] for row in state]
        if direction == 'up' and i > 0:
            new_state[i][j], new_state[i - 1][j] = (new_state[i - 1][j], new_state[i][j])
        elif direction == 'down' and i < 2:
            new_state[i][j], new_state[i + 1][j] = (new_state[i + 1][j], new_state[i][j])
        elif direction == 'left' and j > 0:
            new_state[i][j], new_state[i][j - 1] = (new_state[i][j - 1], new_state[i][j])
        elif direction == 'right' and j < 2:
            new_state[i][j], new_state[i][j + 1] = (new_state[i][j + 1], new_state[i][j])
        return new_state

    def get_possible_moves(self, state):
        i, j = self.find_blank(state)
        moves = []
        if i > 0:
            moves.append('up')
        if i < 2:
            moves.append('down')
        if j > 0:
            moves.append('left')
        if j < 2:
            moves.append('right')
        return moves

    def stateToString(self, state):
        return ''.join((str(num) for row in state for num in row))

    def solve(self):
        start = tuple((tuple(row) for row in self.initial_state))
        goal = tuple((tuple(row) for row in self.goal_state))
        if start == goal:
            return []
        open_list = collections.deque()
        closed_set = set()
        open_list.append((start, []))
        while open_list:
            state, path = open_list.popleft()
            state_str = self.stateToString(state)
            if state_str in closed_set:
                continue
            closed_set.add(state_str)
            if state == goal:
                return path
            for move in self.get_possible_moves(state):
                new_state = self.move(state, move)
                new_state_tuple = tuple((tuple(row) for row in new_state))
                if self.stateToString(new_state_tuple) not in closed_set:
                    open_list.append((new_state_tuple, path + [move]))
        return None

class Node:
    pass

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_cpp/EightPuzzle.cpp
#include <array>
#include <vector>
#include <list>
#include <queue>
#include <unordered_set>
#include <string>
#include <stdexcept>
#include <utility>

using State = std::array<std::array<int, 3>, 3>;

class EightPuzzle {
private:
    State initialState;
    State goalState;

public:
    EightPuzzle(const State& initialState) : initialState(initialState) {
        goalState = { { {1,2,3}, {4,5,6}, {7,8,0} } };
    }

    std::pair<int, int> findBlank(const State& state) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (state[i][j] == 0) {
                    return std::make_pair(i, j);
                }
            }
        }
        throw std::runtime_error("Blank not found");
    }

    State move(const State& state, const std::string& direction) {
        std::pair<int, int> blank = findBlank(state);
        int i = blank.first;
        int j = blank.second;
        State newState = state;

        if (direction == "up") {
            if (i > 0) {
                newState[i][j] = newState[i-1][j];
                newState[i-1][j] = 0;
            }
        } else if (direction == "down") {
            if (i < 2) {
                newState[i][j] = newState[i+1][j];
                newState[i+1][j] = 0;
            }
        } else if (direction == "left") {
            if (j > 0) {
                newState[i][j] = newState[i][j-1];
                newState[i][j-1] = 0;
            }
        } else if (direction == "right") {
            if (j < 2) {
                newState[i][j] = newState[i][j+1];
                newState[i][j+1] = 0;
            }
        }
        return newState;
    }

    std::vector<std::string> getPossibleMoves(const State& state) {
        std::pair<int, int> blank = findBlank(state);
        int i = blank.first;
        int j = blank.second;
        std::vector<std::string> moves;

        if (i > 0) moves.push_back("up");
        if (i < 2) moves.push_back("down");
        if (j > 0) moves.push_back("left");
        if (j < 2) moves.push_back("right");

        return moves;
    }

    std::list<std::string>* solve() {
        struct Node {
            State state;
            std::list<std::string> path;
            Node(const State& s, const std::list<std::string>& p) : state(s), path(p) {}
        };

        std::queue<Node> openList;
        std::unordered_set<std::string> closedList;

        openList.push(Node(initialState, std::list<std::string>()));

        while (!openList.empty()) {
            Node currentNode = openList.front();
            openList.pop();
            State currentState = currentNode.state;
            std::string currentStateString = stateToString(currentState);

            if (closedList.find(currentStateString) != closedList.end()) {
                continue;
            }
            closedList.insert(currentStateString);

            if (currentState == goalState) {
                return new std::list<std::string>(currentNode.path);
            }

            std::vector<std::string> possibleMoves = getPossibleMoves(currentState);
            for (const auto& move : possibleMoves) {
                State newState = move(currentState, move);
                std::string newStateString = stateToString(newState);
                if (closedList.find(newStateString) == closedList.end()) {
                    std::list<std::string> newPath = currentNode.path;
                    newPath.push_back(move);
                    openList.push(Node(newState, newPath));
                }
            }
        }

        return nullptr;
    }

private:
    std::string stateToString(const State& state) {
        std::string s;
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                s += ('0' + state[i][j]);
            }
        }
        return s;
    }
};